### PR TITLE
Update StackedTransformer to allow input data modification

### PR DESCRIPTION
**Pull request by qdavid1**

### CODE REVIEW

File: axlearn/common/attention.py [hunk 1]
```diff
@@ -3589,6 +3590,30 @@ def _forward_for_mode(
...
+        This method is called before we invoke each layer in `self._layers`.
+        The updated `data` will be passed to the layer invocation.
+
+        Args:
+            data: a Tensor denoting the input data to the upcoming layer.
+            all_layer_outputs: a list of BaseTransformerLayer.Output that is appended with
+                the output of each constituent layer in the stack.
+
+        Returns: a new Tensor.
+
```
Comment (ruomingp) [inline axlearn/common/attention.py:3611]: Fix the comment format?

File: axlearn/common/attention.py [hunk 2]
```diff
@@ -3563,6 +3563,7 @@ def _forward_for_mode(
         all_layer_states = []
         for i, layer in enumerate(self._layers):
             # Prepare inputs to the current layer.
+            data = self._update_data(data, all_layer_outputs=all_layer_outputs)
```
Comment (ruomingp) [inline axlearn/common/attention.py:3566]: Add a test for this logic?

**Review by ruomingp — COMMENTED** [review]
Thanks.

File: axlearn/common/attention.py [hunk 3]
```diff
@@ -3589,6 +3590,30 @@ def _forward_for_mode(
 
         return all_layer_states, self._aggregate_layer_outputs(all_layer_outputs)
 
+    # pylint: disable=unused-argument
```
Comment (markblee) [inline axlearn/common/attention.py:3593]: Could we use `del all_layer_outputs` in function body or `disable-next`?

File: axlearn/common/attention.py [hunk 4]
```diff
@@ -3589,6 +3590,30 @@ def _forward_for_mode(
...
+        This method is called before we invoke each layer in `self._layers`.
+        The updated `data` will be passed to the layer invocation.
+
+        Args:
+            data: a Tensor denoting the input data to the upcoming layer.
+            all_layer_outputs: a list of BaseTransformerLayer.Output that is appended with
+                the output of each constituent layer in the stack.
+
+        Returns:
+            a new Tensor.
```
Comment (markblee) [inline axlearn/common/attention.py:3611]:
```suggestion
        Args:
            data: A Tensor denoting the input data to the upcoming layer.
            all_layer_outputs: A list of BaseTransformerLayer.Output that is appended with
                the output of each constituent layer in the stack.

        Returns:
            A new Tensor.
```

**Comment by qdavid1** [comment]
Thanks for the reviews! I updated the code to address the four suggestions.

**Review by markblee — APPROVED** [review]
As a reminder, please comment "done" when reviewer comments have been addressed. Thanks!

File: axlearn/common/attention.py (hunk 4 above)
Comment (qdavid1) [reply axlearn/common/attention.py]: Done

File: axlearn/common/attention.py (hunk 3 above)
Comment (qdavid1) [reply axlearn/common/attention.py]: Done

File: axlearn/common/attention.py (hunk 2 above)
Comment (qdavid1) [reply axlearn/common/attention.py:3566]: Done

File: axlearn/common/attention.py (hunk 1 above)
Comment (qdavid1) [reply axlearn/common/attention.py]: Done

**Review by ruomingp — APPROVED** [review]

**Pull request merged**